# Day 4: Data Analysis with Pandas
**Internship Program - Week 1**

### Objectives:
1. Master core Pandas structures: `Series` and `DataFrame`.
2. Load and inspect tabular data efficiently.
3. Perform data selection and conditional filtering (`.loc`, `.iloc`).
4. Handle missing data and duplicate records.
5. Aggregate and group data using `groupby` and `agg`.

## Import necessary libraries

In [1]:
import pandas as pd
import numpy as np

print(f"Pandas Version: {pd.__version__}")
print(f"NumPy Version: {np.__version__}")

Pandas Version: 3.0.3
NumPy Version: 2.5.1


## 1. Core Pandas Data Structures

In [4]:
# create pandas series
data_series = pd.Series([100,200, 300, 400], index =['Q1', 'Q2', 'Q3', 'Q4'], name="Revenue")
print("--- Pandas Series ---")
print(data_series)
print("\n")

# create a dataFrame from dictionary
raw_data = {
    'Employee_ID': [101, 102, 103, 104, 105, 106, 107],
    'Name': ['Alice', 'Bob', 'Charlie', 'David', 'Eva', 'Frank', 'Grace'],
    'Department': ['HR', 'IT', 'IT', 'Finance', 'HR', 'Finance', 'IT'],
    'Age': [25, 30, 35, np.nan, 22, 40, 29],
    'Salary': [50000, 75000, 82000, 65000, np.nan, 90000, 72000],
    'Experience_Years': [2, 5, 8, 4, 1, 12, np.nan]
}
df= pd.DataFrame(raw_data)
print("--- DataFrame ---")
df

--- Pandas Series ---
Q1    100
Q2    200
Q3    300
Q4    400
Name: Revenue, dtype: int64


--- DataFrame ---


,Employee_ID,Name,Department,Age,Salary,Experience_Years
0,101,Alice,HR,25.0,50000.0,2.0
1,102,Bob,IT,30.0,75000.0,5.0
2,103,Charlie,IT,35.0,82000.0,8.0
3,104,David,Finance,NaN,65000.0,4.0
4,105,Eva,HR,22.0,NaN,1.0
5,106,Frank,Finance,40.0,90000.0,12.0
6,107,Grace,IT,29.0,72000.0,NaN


## 2. Data Inspection & Overview

Before analyzing data, we inspect its shape, column types, missing values, and statistical properties using built-in methods like `.head()`, `.info()`, and `.describe()`.

In [7]:
# Display first 5 rows
print("--- First 5 Rows ---")
print(df.head())

# Display DataFrame information (column types, non-null counts)
print("\n--- DataFrame Information ---")
print(df.info())

# Summary statistics for numerical columns
print("\n--- Descriptive Statistics ---")
print(df.describe())

--- First 5 Rows ---
   Employee_ID     Name Department   Age   Salary  Experience_Years
0          101    Alice         HR  25.0  50000.0               2.0
1          102      Bob         IT  30.0  75000.0               5.0
2          103  Charlie         IT  35.0  82000.0               8.0
3          104    David    Finance   NaN  65000.0               4.0
4          105      Eva         HR  22.0      NaN               1.0

--- DataFrame Information ---
<class 'pandas.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Employee_ID       7 non-null      int64  
 1   Name              7 non-null      str    
 2   Department        7 non-null      str    
 3   Age               6 non-null      float64
 4   Salary            6 non-null      float64
 5   Experience_Years  6 non-null      float64
dtypes: float64(3), int64(1), str(2)
memory usage: 468.0 bytes
None

--- De

## 3. Data Selection & Boolean Filtering

Pandas provides powerful indexing tools:
- **Column Selection**: Accessing single or multiple columns.
- **`.loc`**: Label-based indexing (selects rows and columns by labels or boolean conditions).
- **`.iloc`**: Integer-position based indexing (selects by numeric position).
- **Boolean Masking**: Filtering rows based on logical conditions.
- 

In [10]:
# 1. Selecting specific columns
names_and_salaries = df[['Name', 'Salary']]
print("--- Selected Columns ---")
print(names_and_salaries.head(3))

# 2. Filtering rows where Salary > 70,000 using Boolean Indexing
high_earners = df[df['Salary'] > 70000]
print("\n--- High Earners (Salary > 70,000) ---")
print(high_earners)

# 3. Using .loc with multiple conditions (IT department AND Salary > 70,000)
it_high_earners = df.loc[(df['Department'] == 'IT') & (df['Salary'] > 70000), ['Name', 'Department', 'Salary']]
print("\n--- IT High Earners via .loc ---")
print(it_high_earners)

# 4. Using .iloc to slice the first 3 rows and first 4 columns
sliced_data = df.iloc[0:3, 0:4]
print("\n--- Sliced Data via .iloc ---")
print(sliced_data)

--- Selected Columns ---
      Name   Salary
0    Alice  50000.0
1      Bob  75000.0
2  Charlie  82000.0

--- High Earners (Salary > 70,000) ---
   Employee_ID     Name Department   Age   Salary  Experience_Years
1          102      Bob         IT  30.0  75000.0               5.0
2          103  Charlie         IT  35.0  82000.0               8.0
5          106    Frank    Finance  40.0  90000.0              12.0
6          107    Grace         IT  29.0  72000.0               NaN

--- IT High Earners via .loc ---
      Name Department   Salary
1      Bob         IT  75000.0
2  Charlie         IT  82000.0
6    Grace         IT  72000.0

--- Sliced Data via .iloc ---
   Employee_ID     Name Department   Age
0          101    Alice         HR  25.0
1          102      Bob         IT  30.0
2          103  Charlie         IT  35.0


## 4. Data Cleaning & Handling Missing Values

Data cleaning involves:
1. Detecting missing values (`NaN`).
2. Imputing missing values using techniques like mean/median filling.
3. Dropping unrecoverable null values.
4. Removing duplicate records.

In [13]:
# Check count of missing values per column
print("--- Missing Values Count Before Cleaning ---")
print(df.isnull().sum())

# Make a copy of the original dataframe for cleaning
df_clean = df.copy()

# 1. Fill missing 'Age' values with the median age (without using inplace=True)
median_age = df_clean['Age'].median()
df_clean['Age'] = df_clean['Age'].fillna(median_age)

# 2. Fill missing 'Salary' values with the mean salary of their respective department
# If a department has all NaNs or single NaN, fill remaining with overall mean
df_clean['Salary'] = df_clean.groupby('Department')['Salary'].transform(lambda x: x.fillna(x.mean()))
df_clean['Salary'] = df_clean['Salary'].fillna(df_clean['Salary'].mean())

# 3. Fill missing 'Experience_Years' with 0
df_clean['Experience_Years'] = df_clean['Experience_Years'].fillna(0)

print("\n--- Cleaned DataFrame (No Missing Values) ---")
print(df_clean)

print("\n--- Missing Values Count After Cleaning ---")
print(df_clean.isnull().sum())

--- Missing Values Count Before Cleaning ---
Employee_ID         0
Name                0
Department          0
Age                 1
Salary              1
Experience_Years    1
dtype: int64

--- Cleaned DataFrame (No Missing Values) ---
   Employee_ID     Name Department   Age   Salary  Experience_Years
0          101    Alice         HR  25.0  50000.0               2.0
1          102      Bob         IT  30.0  75000.0               5.0
2          103  Charlie         IT  35.0  82000.0               8.0
3          104    David    Finance  29.5  65000.0               4.0
4          105      Eva         HR  22.0  50000.0               1.0
5          106    Frank    Finance  40.0  90000.0              12.0
6          107    Grace         IT  29.0  72000.0               0.0

--- Missing Values Count After Cleaning ---
Employee_ID         0
Name                0
Department          0
Age                 0
Salary              0
Experience_Years    0
dtype: int64


## 5. Data Grouping & Aggregation

The `groupby` operation splits the dataset into groups based on specified criteria, applies functions (like sum, mean, count) to each group, and combines the results.


In [16]:
# 1.Average salary per dept
avg_salary_per_dept= df_clean.groupby('Department')['Salary'].mean().reset_index()
print("--- Average salary per dept --- ")
print(avg_salary_per_dept)

# 2.Multiple statistics per group
dept_summary = df_clean.groupby('Department').agg(
    Employee_count=('Employee_ID', 'count'),
    Avg_salary= ('Salary', 'mean'),
    Max_Salary= ('Salary','max'),
    Avg_Experience= ('Experience_Years','mean')
).reset_index()

print("\n--- Detailed Department Summary ---")
print(dept_summary)

--- Average salary per dept --- 
  Department        Salary
0    Finance  77500.000000
1         HR  50000.000000
2         IT  76333.333333

--- Detailed Department Summary ---
  Department  Employee_count    Avg_salary  Max_Salary  Avg_Experience
0    Finance               2  77500.000000     90000.0        8.000000
1         HR               2  50000.000000     50000.0        1.500000
2         IT               3  76333.333333     82000.0        4.333333


## 6. Saving Cleaned Data & Summary

In [18]:
# Save clean dataset to CSV inside Day_4 directory
df_clean.to_csv('cleaned_employee_data.csv', index=False)
print("Cleaned data successfully saved to 'cleaned_employee_data.csv'.")

Cleaned data successfully saved to 'cleaned_employee_data.csv'.


# Hands-On Lab: Exploring a Real Dataset with Pandas

## Overview
In this practical lab, we perform an end-to-end exploratory data analysis (EDA) on a real-world supermarket sales dataset directly loaded via URL. The workflow covers:
1. **Dataset Loading & Inspection:** Loading CSV data and reporting shape, columns, and data types.
2. **Data Cleaning & Missing Value Handling:** Identifying missing values and applying justified imputation or dropping strategies.
3. **Subsetting & Filtering:** Extracting meaningful slices of data based on specific conditions.
4. **Group Aggregation:** Computing descriptive metrics grouped by categorical features using `groupby`.

In [8]:
#Official raw CSV dataset from GitHub (Seaborn Official Repository)
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/mpg.csv"

# Load dataset into Pandas DataFrame
df_lab = pd.read_csv(url)

print("✅ Dataset successfully loaded into DataFrame!\n")

# Report dataset shape
print(f"1. Dataset Shape (Rows, Columns): {df_lab.shape}\n")

# Report list of columns and data types
print("2. Column Names and Data Types:")
print(df_lab.dtypes)

# Preview first 5 rows
print("\n3. First 5 Rows Preview:")
df_lab.head()

✅ Dataset successfully loaded into DataFrame!

1. Dataset Shape (Rows, Columns): (398, 9)

2. Column Names and Data Types:
mpg             float64
cylinders         int64
displacement    float64
horsepower      float64
weight            int64
acceleration    float64
model_year        int64
origin              str
name                str
dtype: object

3. First 5 Rows Preview:


,mpg,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,name
0,18.0,8,307.0,130.0,3504,12.0,70,usa,chevrolet chevelle malibu
1,15.0,8,350.0,165.0,3693,11.5,70,usa,buick skylark 320
2,18.0,8,318.0,150.0,3436,11.0,70,usa,plymouth satellite
3,16.0,8,304.0,150.0,3433,12.0,70,usa,amc rebel sst
4,17.0,8,302.0,140.0,3449,10.5,70,usa,ford torino


## Step 2: Handling Missing Values & Data Cleaning

### Documented Justification:
1. **Detection:** We count null values across all columns to locate missing data points.
2. **Imputation Strategy:**
   - Numerical columns with missing values (e.g., `horsepower`) are filled using the **median** value of that column.
   - *Why Median?* The median is robust against extreme outliers, ensuring we do not distort the central tendency of engine power metrics.
3. **Deduplication:** Duplicate rows are checked and dropped to preserve dataset integrity.

In [9]:
# 1. Count missing values per column
print("--- Missing Values per Column before Cleaning---")
print(df_lab.isnull().sum())

# 2. Impute missing values in numeric columns using median
numeric_cols = df_lab.select_dtypes(include=['float64', 'int64']).columns
for col in numeric_cols:
    if df_lab[col].isnull().sum() > 0:
        median_val = df_lab[col].median()
        df_lab[col] = df_lab[col].fillna(median_val)
        print(f"\n-> Imputed missing values in '{col}' with median: {median_val}")

# 3. check and drop duplicate rows
initial_rows = len(df_lab)
df_lab.drop_duplicates(inplace=True)

print(f"\nDuplicates removed: {initial_rows - len(df_lab)}")
print(f"Total missing values remaining: {df_lab.isnull().sum().sum()}")

--- Missing Values per Column before Cleaning---
mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model_year      0
origin          0
name            0
dtype: int64

-> Imputed missing values in 'horsepower' with median: 93.5

Duplicates removed: 0
Total missing values remaining: 0


## Step 3: Data Filtering & Subset Exploration

### Subset Description & Findings:
- **Filtering Threshold:** We filter the dataset to isolate high fuel-efficiency cars where Miles Per Gallon (`mpg`) exceeds **30.0**.
- **Analytical Goal:** Explore the physical properties (cylinders, weight, horsepower) of eco-friendly vehicles compared to the general population.

In [10]:
# filter rows where mpg > 30.0
high_efficiency_cars = df_lab[df_lab['mpg'] > 30.0]

total_count = len(df_lab)
subset_count = len(high_efficiency_cars)
percentage = (subset_count / total_count) * 100

print(f"Total Car Models: {total_count}")
print(f"High Fuel-Efficiency Models (mpg > 30): {subset_count}")
print(f"Percentage of High Efficiency Models: {percentage:.2f}%\n")

# Preview specific attributes of the subset
display_cols = ['name', 'mpg', 'cylinders', 'horsepower', 'weight', 'origin']
print("--- High Fuel-Efficiency Sample Preview ---")
high_efficiency_cars[display_cols].head()

Total Car Models: 398
High Fuel-Efficiency Models (mpg > 30): 85
Percentage of High Efficiency Models: 21.36%

--- High Fuel-Efficiency Sample Preview ---


,name,mpg,cylinders,horsepower,weight,origin
53,toyota corolla 1200,31.0,4,65.0,1773,japan
54,datsun 1200,35.0,4,69.0,1613,japan
129,datsun b210,31.0,4,67.0,1950,japan
131,toyota corolla 1200,32.0,4,65.0,1836,japan
144,toyota corona,31.0,4,52.0,1649,japan


## Step 4: Groupby Categorical Aggregation

### Aggregation Metrics:
We group the dataset by manufacturing region (`origin`) to calculate:
1. **`Average_MPG`**: Mean fuel economy per origin.
2. **`Average_Horsepower`**: Average engine output.
3. **`Average_Weight`**: Average vehicle mass.
4. **`Model_Count`**: Number of car models tested from each region.

In [11]:
# Aggregate stats by manufacturing origin
origin_summary = df_lab.groupby('origin').agg(
    Average_MPG=('mpg', 'mean'),
    Average_Horsepower=('horsepower', 'mean'),
    Average_Weight=('weight', 'mean'),
    Model_Count=('name', 'count')
).reset_index()

# Sort by Average MPG descending
origin_summary = origin_summary.sort_values(by='Average_MPG', ascending=False)

print("--- Regional Vehicle Performance Summary ---")
origin_summary

--- Regional Vehicle Performance Summary ---


,origin,Average_MPG,Average_Horsepower,Average_Weight,Model_Count
1,japan,30.450633,79.835443,2221.227848,79
0,europe,27.891429,80.928571,2423.300000,70
2,usa,20.083534,118.638554,3361.931727,249


##  Final Interpretation of Results

Based on the aggregated metrics grouped by manufacturing region (`origin`), we draw the following conclusions:

1. **Fuel Efficiency Leadership:**
   - **Japan** leads in fuel economy with an average of **30.45 MPG**, closely followed by **Europe** at **27.89 MPG**.
   - **USA** lags behind with an average of **20.08 MPG**.

2. **Power & Weight Trade-off:**
   - **American vehicles** prioritize high performance and size, having the highest average horsepower (**118.64 HP**) and average weight (**3361.93 lbs**).
   - **Japanese and European vehicles** feature lighter builds (~2221 lbs and ~2423 lbs) and moderate power output (~80 HP), which directly contributes to their superior fuel efficiency.

3. **Sample Distribution:**
   - The dataset is predominantly composed of American models (**249 models**), compared to Japanese (**79 models**) and European (**70 models**).

---
###  Summary Conclusion
Lighter vehicle weight and lower engine power output are strong indicators of high fuel efficiency (`MPG`). Japanese manufacturers achieved the best balance between weight reduction and fuel performance in this dataset.